pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,490 @@
+    \end{itemize}
+    \inRust{usize}{https://doc.rust-lang.org/std/primitive.usize.html}
+\end{definition}
+
+\begin{definition}[\texttt{Vec(T)}]$\sfi$
+    The rust type \texttt{Vec(T)} consists of ordered lists of type \texttt{T}. For example, if \texttt{T = bool}, then values of type \texttt{Vec(T)} include $\texttt{[], [0], [1], [0, 0],} \ldots$ \inRust{Vec<T>}{https://doc.rust-lang.org/std/vec/struct.Vec.html}
+\end{definition}
+
+\horizline
+
+\subsection{Notes, todos, questions}
+
+%\silvia{Discuss: bounded vs lower bound and upper bound}
+
+\todonei{Include info on MPFR, and possibly relate it to our existing definitions of floats.}
+
+\todonei{Define plus, minus, etc. below each type on which they operate. For example, the definition for \texttt{u32} should also include a definition of plus on \texttt{u32}, multiplication on \texttt{u32}, etc.}
+
+\questionr{Should we have a general definition for ``floats'' (and ``integers''?), or is it sufficiently understood what a float is in general?}
+
+\section{Domains}
+
+$\sfi$ A \emph{data domain} is a representation of the set of values on which a metric or function can operate. For example, if a function accepts inputs from the domain \texttt{IntervalDomain(1:u32,17:u32)}, this means that the function can take any input value \texttt{v} of type \texttt{u32} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\questionr{Should anything else be included in this introduction to ``Domains''?}
+
+\begin{definition}[\texttt{AllDomain}]$\sfi$
+\texttt{AllDomain(T)} is the data domain consisting of all values of type $\T$. 
+\end{definition}
+
+For example, \texttt{AllDomain(u32)} is the domain of all values of type \texttt{u32}.
+
+\begin{definition}[\texttt{IntervalDomain}] \sfi
+For a type \texttt{T} that has a total ordering (\texttt{T} has trait \texttt{TotalOrd}) and for values \texttt{L <= U} of type \texttt{T}, \texttt{IntervalDomain(L:T, U:T)} is the domain that contains all values \texttt{v} of type \texttt{T} such that \texttt{L <= v} and \texttt{v <= U}.
+    
+Note that, because both \texttt{L} and \texttt{U} are of type \texttt{T}, there is no need to explicitly pass \texttt{T}; the type \texttt{T} can be inferred. \texttt{IntervalDomain} is defined on any type that implements the trait \texttt{TotalOrd}.\footnote{As of June 28, the OpenDP library requires the weaker condition of partial ordering (implements \texttt{PartialOrd}) instead.}
+\end{definition}
+
+For example, \texttt{IntervalDomain(1:u32, 17:u32)} corresponds to a domain that contains all the \texttt{u32} values \texttt{v} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\begin{definition}[\texttt{SizedDomain}] $\sfi$
+    \texttt{SizedDomain(D)} is a domain that specifies the length of the enclosed domain \texttt{D}. The length of elements of the enclosed domain is specified with an element of type \texttt{usize}.
```

Comment:
Shouldn't it be "SizedDomain(D,n)" rther than SizedDomain(D)?  This should also appear after the simpler definition of VectorDomain. And should be written similarly to vector domain "SizedDomain(D,n) is the domain of all vectors of length n drawn from Domain D"